src/provider.tsx

```diff
@@ -0,0 +1,160 @@
+  options?: PopupLoginOptions,
+  config?: PopupConfigOptions,
+): Promise<void> => {
+  dispatch(loginWithPopupAction.started());
+
+  try {
+    await auth0Client.loginWithPopup(options, config);
+  } catch (error) {
+    console.error(error);
+  }
+
+  const user = await auth0Client.getUser();
+
+  dispatch(loginWithPopupAction.done({ result: { user } }));
+};
+
+export const Auth0Provider: React.FC<Auth0ProviderProps> = ({
+  children,
+  history,
+  enableDebugLogging,
+  renderLoading,
+  ...initOptions
+}) => {
+  const log = useLog(enableDebugLogging);
+
+  const [state, dispatch] = useReducer<Reducer<Auth0ProviderState, Auth0ProviderAction>>(
+    auth0ProviderStateReducer,
+    InitialReducerState,
+  );
+
+  const location = useLocation();
+
+  const { loading: initLoading, value: initValue } = useAsync(async () => {
+    log("Beginning auth0 initialization");
+
+    dispatch(auth0Loaded.started());
+
+    const auth0Client = await createAuth0Client(initOptions);
+
+    const handleRedirectCallback = buildHandleRedirectCallback(dispatch, log, auth0Client);
+
+    const loginWithPopup = buildLoginWithPopup(dispatch, auth0Client);
+
+    // if (location.search?.includes("code=") && location.search?.includes("state=")) {
+    //   log("Handling redirect");
+    //   const { appState } = await handleRedirectCallback();
+    //   onRedirectCallback(appState);
+    // }
+
+    const isAuthenticated = await auth0Client.isAuthenticated();
+
+    const user: TokenUser | undefined = (isAuthenticated && (await auth0Client.getUser())) || undefined;
+
+    return {
+      user,
+      auth0Client,
+      handleRedirectCallback,
+      loginWithPopup,
+    };
+  }, []);
+
+  const [handleRedirAsyncState, handleRedirAsync] = useAsyncFn(async () => {
+    if (!isLoadedState(state)) {
+      return;
+    }
+
+    log("Handling redirect");
+    const { appState: s } = await state.handleRedirectCallback();
+    const appState: RouteState = s;
+    log("Back from redirect, replacing state to ", appState);
+    history.replace(appState.targetUrl);
+  }, ["handleRedirectCallback" in state && state.handleRedirectCallback, log, history.replace]);
+
+  useEffect(() => {
+    if (!isLoadedState(state) || handleRedirAsyncState.loading) {
+      return;
+    }
+
+    if (location.search?.includes("code=") && location.search?.includes("state=")) {
+      handleRedirAsync().then(() => {
+        /* no-op */
+      });
+    }
+  }, [location.search, state, handleRedirAsyncState.loading, handleRedirAsync]);
+
+  useEffect(() => {
+    if (isLoadingState(state) && !initLoading && initValue) {
+      log("auth0 initialization complete");
+      dispatch(auth0Loaded.done({ result: initValue }));
+    }
+  }, [state, dispatch, initLoading, log, initValue]);
+
+  if (isLoadingState(state)) {
+    log("Rendering loading view due to auth0Client still loading");
+    return (renderLoading && renderLoading(state)) || <></>;
```

Comment:
may not matter but thinking about rendering an empty fragment vs `null`

```
return (renderLoading && renderLoading(state));
```

Reply:
Interesting.  I had `null` originally, and I could _swear_ it broke the return type of `React.FC`.  Of course, I'm a dirty liar, and that's not true.  I prefer `null` as well and am putting it back in.